Make competitive iterators more robust.

jpountz opened the pull request:
As per a recent bug (#14517), competitive iterators are hard to get right given how their state gets updated in place. This commit tries to make them more robust by extracting the logic of updating the state of a `DocIdSetIterator` to a shared class, which can then be tested on its own.

A side-effect is that it implements `#intoBitSet` on the doc competitive iterator and `#docIDRunEnd` on all competitive iterators.

jpountz committed:
Make competitive iterators more robust.

As per a recent bug (#14517), competitive iterators are hard to get
right given how their state gets updated in place. This commit tries to
make them more robust by extracting the logic of updating the state of a
`DocIdSetIterator` to a shared class, which can then be tested on its
own.

A side-effect is that it implements `#intoBitSet` on the doc competitive
iterator and `#docIDRunEnd` on all competitive iterators.

jpountz committed:
Remove unused variable

jpountz committed:
fix test failure

jpountz committed:
Merge branch 'main' into robust_competitive_iterators

gf2121 reviewed (approved):
I like the simplification,

> A side-effect is that it implements #intoBitSet on the doc competitive iterator and #docIDRunEnd on all competitive iterators.

By 'side-effect', do you mean another gain? I thought it means negative effects :)

jpountz commented:
> By 'side-effect', do you mean another gain? I thought it means negative effects :)

Hopefully a gain, I still need to run benchmarks. This change also increases polymorphism, so I'm not sure if it will make things better or not.

jpountz commented:
I could confirm that it's due to `#docIdEnd()` now being implemented on the competitive iterators. This triggers usage of `DISIDocIdStream`, which calls `nextDoc()` in a loop, and calling `nextDoc()` in a loop happens to be a slower way of iterating doc IDs than calling `intoBitSet` and then iterating set bits when blocks are encoded as a bit set.

jpountz committed:
Remove DISIDocIdStream.

I had initially introduced `DISIDocIdStream` to avoid introducing regressions
when `DenseConjunctionBulkScorer` started accepting single clauses. However,
benchmarks on #14532 suggested that going through `DISIDocIdStream` is slower
than loading docs into a bit set first and then iterating the bit set, when the
postings list has many of its blocks encoded as bit sets.

This makes sense, the way how `BitSetDocIdStream` iterates set bits saves a
number of operations compared with calling `FixedBitSet#nextSetBit` in a loop.

So I'm suggesting removing `DISIDocIdStream` for now for simplicity.

jpountz committed:
Merge branch 'main' into robust_competitive_iterators
